huxley/invoice_automation/src/module/quick_books_module.py

```diff
@@ -0,0 +1,383 @@
+SCHOOL_FEE = "School Fee"
+DELEGATE_FEE = "Delegate Fee"
+CC_FEE_ITEM_NAME = "Credit Card Processing Fee"
+
+
+class QuickBooksModule:
+    """
+    Abstraction barrier for calling QuickBook API's
+
+    Attributes
+    ----------
+    auth_client: AuthClient
+        OAuth2 authentication client for authenticating to QuickBooks API
+    quickbooks_client: QuickBooks
+        QuickBooksClient which makes actual API calls to QuickBooks
+
+    Methods
+    -------
+    querySchoolsAsCustomers(schoolNames: List[str]) -> List[School]:
+        Looks for already registered schools using their names
+    """
+    REDIRECT_URI = "http://localhost:8000"
+    ENVIRONMENT = "production"
+
+    def __init__(
+            self,
+            client_id: str,
+            client_secret: str,
+            refresh_token="",
+            realm_id="",
+            access_token=""
+    ) -> None:
+        """
+        Instantiates authClient using hardcoded CLIENT_ID and CLIENT_SECRET
+        then uses it to instantiate a QuickBooks instance
+
+        Raises
+        ------
+        QuickbooksException
+        """
+
+        self.auth_client = AuthClient(
+            client_id=client_id,
+            client_secret=client_secret,
+            redirect_uri=self.REDIRECT_URI,
+            environment=self.ENVIRONMENT,
+            access_token=access_token
+        )
+        self.auth_client.refresh_token = refresh_token
+
+        self.quickbooks_client = QuickBooks(
+            auth_client=self.auth_client,
+            refresh_token=refresh_token,
+            company_id=realm_id,
+        )
+        # Check connection, refresh if needed
+        self.try_with_refresh(Customer.all, qb=self.quickbooks_client)
+
+        self.credit_card_processing_fee = None
+
+    def try_with_refresh(self, func, *args, **kwargs):
+        """
+        Tries calling func, a quickbooks api call
+        If it fails due to an AuthorizationException, tries refreshing the access_token and then retries
+        :param func:
+        :param args:
+        :param kwargs:
+        :return:
+        """
+        try:
+            try:
+                return func(*args, **kwargs)
+            except AuthorizationException:
+                self.auth_client.refresh()
+                return func(*args, **kwargs)
+        except QuickbooksException as e:
+            print(e.message)
+            print(e.error_code)
+            print(e.detail)
+            raise e
+
+    # Customer methods:
+
+    def query_schools_as_customers(self, school_names: List[str]) -> List[School]:
+        """
+        Looks for existing customers with the same display names as the passed school names
+
+        :param school_names: List of strings containing school names
+        :return: List of school objects corresponding to customers which were found
+        :raises QuickbooksException:
+        """
+        # qbCustomers = Customer.choose(school_names, DISPLAY_NAME, self.quickbooks_client)
```

Comment:
can this commented out be removed? similar for other commented out code noted below in this fine

Reply:
I included this and other commented out code for readability, since it might not be immediately obvious how the syntax of the retry wrapper translates to the actual function call. Would you still want to remove? I can also add something like `# try with retry:` or something similar to make it clear that the comment is showing what the function call w/o retry would look like